### PR TITLE
[improvement](memory) Support GC segment cache, when memory insufficient

### PR DESCRIPTION
**Pull request by xinyiZzz**
# Proposed changes

Issue Number: close #xxx

## Problem summary

1. fix segment cache memory tracker statistics
2. support GC

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.98 seconds
 stream load tsv:          461 seconds loaded 74807831229 Bytes, about 154 MB/s
 stream load json:         38 seconds loaded 2358488459 Bytes, about 59 MB/s
 stream load orc:          67 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          27 seconds loaded 861443392 Bytes, about 30 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230221111840_clickbench_pr_101045.html

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/util/mem_info.cpp [hunk 1]
```diff
@@ -97,6 +98,8 @@ void MemInfo::process_cache_gc(int64_t& freed_mem) {
     freed_mem +=
             StoragePageCache::instance()->get_page_cache_mem_consumption(segment_v2::DATA_PAGE);
     StoragePageCache::instance()->prune(segment_v2::DATA_PAGE);
+    freed_mem += SegmentLoader::instance()->segment_cache_mem_consumption();
+    SegmentLoader::instance()->prune_all();
```
Comment (zhannngchen) [inline be/src/util/mem_info.cpp:102]: It's too aggressive, we may just need to call `prune_all()` when full gc.
Segment cache is quite useful on accelerating queries and MoW index lookup.

Comment (xinyiZzz) [reply be/src/util/mem_info.cpp]: fixed,
minor gc only prune outdated segment cache
full gc prune all

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run arm

File: be/src/util/mem_info.cpp [hunk 2]
```diff
@@ -97,6 +98,8 @@ void MemInfo::process_cache_gc(int64_t& freed_mem) {
     freed_mem +=
             StoragePageCache::instance()->get_page_cache_mem_consumption(segment_v2::DATA_PAGE);
     StoragePageCache::instance()->prune(segment_v2::DATA_PAGE);
+    // TODO add freed_mem
+    SegmentLoader::instance()->prune();
```
Comment (zhannngchen) [inline be/src/util/mem_info.cpp:102]: Why do we need to call `prune()` here?

Comment (zhannngchen) [reply be/src/util/mem_info.cpp:102]: with the pr: https://github.com/apache/doris/pull/17011, the `prune()` here seems not necessary.

File: be/src/olap/segment_loader.cpp [hunk 3]
```diff
@@ -52,8 +53,15 @@ void SegmentLoader::_insert(const SegmentLoader::CacheKey& key, SegmentLoader::C
...
 
+    int64_t meta_mem_usage = 0;
+    for (auto segment : value.segments) {
+        meta_mem_usage += segment->meta_mem_usage();
+    }
+
     auto lru_handle = _cache->insert(key.encode(), &value, sizeof(SegmentLoader::CacheValue),
-                                     deleter, CachePriority::NORMAL);
+                                     deleter, CachePriority::NORMAL, meta_mem_usage);
+    LOG(INFO) << "1111111 " << sizeof(SegmentLoader::CacheValue) << ", " << meta_mem_usage << ", "
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply be/src/olap/segment_loader.cpp:63]: fixed, mistake……

File: be/src/util/mem_info.cpp (hunk 2 above)
Comment (xinyiZzz) [reply be/src/util/mem_info.cpp:102]: minor gc, the remaining memory is less than 10% of the total memory, so prune outdated segment cache

Comment (xinyiZzz) [reply be/src/util/mem_info.cpp:102]: not conflict with periodic gc in #17011

**Review by zhannngchen — APPROVED** [review]
LGTM

**Comment by xinyiZzz** [comment]
run buildall

**Pull request merged**